wallet/src/Pos/Wallet/Web/Pending/Worker.hs

```diff
@@ -0,0 +1,167 @@
+    let ptxSlot = flattenSlotId (_ptxCreationSlot ptx)
+        checkStartSlot = ptxSlot + initialDelay
+    and [ curSlot > checkStartSlot
+        , ((curSlot - checkStartSlot) `mod` furtherDelay) == 0
+        ]
+  where
+    initialDelay = 3 :: FlatSlotId
+    furtherDelay = 1 :: FlatSlotId
+
+resubmitTx :: MonadPendings m => SendActions m -> PendingTx -> m ()
+resubmitTx SendActions{..} ptx@PendingTx{..} = do
+    logInfo $ sformat ("Resubmitting tx "%build) _ptxTxId
+    -- FIXME [CSM-256] Doesn't it introduce a race condition?
+    void (submitAndSaveTx enqueueMsg _ptxTxAux) `catches` handlers
+  where
+    handlers =
+        [ Handler $ \e -> do
+            reportFail "Failed to resubmit tx" e
+            processPtxFailure ptx e
+
+        , Handler $ \(SomeException e) ->
+            reportFail "Failed to resubmit tx, will try again later" e
+        ]
+    reportFail :: MonadPendings m => Exception e => Text -> e -> m ()
+    reportFail desc =
+        logInfo . sformat (stext%" "%build%": "%shown) desc _ptxTxId
+
+-- | Distributes pending txs submition over current slot ~evenly
+resubmitPtxsDuringSlot
+    :: MonadPendings m
+    => SendActions m -> [PendingTx] -> m ()
+resubmitPtxsDuringSlot sendActions ptxs = do
+    interval <- evalSubmitDelay (length ptxs)
+    forM_ ptxs $ \ptx -> do
+        delay interval
+        fork $ resubmitTx sendActions ptx
+  where
+    submitionEta = 5 :: Second
+    evalSubmitDelay toResubmitNum = do
+        slotDuration <- getLastKnownSlotDuration
+        let checkPeriod = max 0 $ slotDuration - convertUnit submitionEta
+        return (checkPeriod `div` fromIntegral toResubmitNum)
+
+processPtxsToResubmit
+    :: MonadPendings m
+    => SendActions m -> [PendingTx] -> m ()
+processPtxsToResubmit sendActions ptxs = do
+    ptxsPerSlotLimit <- evalPtxsPerSlotLimit
+    let toResubmit =
+            take ptxsPerSlotLimit $
+            filter ((PtxApplying ==) . _ptxCond)
+            ptxs
+    logInfo $ sformat fmt (map _ptxTxId toResubmit)
+    resubmitPtxsDuringSlot sendActions toResubmit
+  where
+    fmt = "Transactions to resubmit on current slot: "%listJson
+    evalPtxsPerSlotLimit = do
+        slotDuration <- getLastKnownSlotDuration
+        let limit = fromIntegral $
+                convertUnit slotDuration `div` pendingTxResubmitionPeriod
+        when (limit <= 0) $
+            logInfo "'pendingTxResubmitionPeriod' is larger than slot duration,\
+                    \ won't resubmit any pending transaction"
+        return limit
+
+-- | Checks and updates state of given pending transactions, resubmitting them
+-- if needed.
+processPtxs
+    :: MonadPendings m
+    => SendActions m -> SlotId -> [PendingTx] -> m ()
+processPtxs sendActions curSlot ptxs = do
+    mapM_ (processPtxInUpperBlocks curSlot) ptxs
+    processPtxsToResubmit sendActions ptxs
+
+processPtxsOnSlot
+    :: MonadPendings m
+    => SendActions m -> SlotId -> m ()
+processPtxsOnSlot sendActions curSlot = do
```

Comment:
This will be reflected on information which transaction history returns.
We store whole transaction history anyway, which is not smaller